File: extension/tools-advanced.js
// Advanced tool handlers — new in CLI Browser Bridge v2.
// wait_for, storage, dom_query

import { cdp, ensureAttached, sleep } from './cdp.js';
import { isInGroup } from './tabs.js';

function notInGroup(tabId) {
  return { content: [{ type: "text", text: `Tab ${tabId} is not in the MCP group.` }] };
}

async function sendContentMessage(tabId, message) {
  try {
    return await chrome.tabs.sendMessage(tabId, message);
  } catch {
    await chrome.scripting.executeScript({ target: { tabId }, files: ["content.js"] });
    return chrome.tabs.sendMessage(tabId, message);
  }
}

export const advancedToolHandlers = {
  async wait_for(args) {
    const { tabId, condition, selector, text, predicate, pierceShadow = true } = args;
    if (!(await isInGroup(tabId))) return notInGroup(tabId);

    const timeout = Math.min(args.timeout || 10, 30) * 1000;
    const pollInterval = 200;
    const start = Date.now();

    while (Date.now() - start < timeout) {
      let met = false;

      switch (condition) {
        case "element_visible": {
          if (!selector) return { content: [{ type: "text", text: "selector required for element_visible" }] };
          const resp = await sendContentMessage(tabId, {
            type: "querySelector", selector, pierceShadow,
          });
          met = resp?.result?.visible === true;
          break;
        }
        case "element_hidden": {
          if (!selector) return { content: [{ type: "text", text: "selector required for element_hidden" }] };
          const resp = await sendContentMessage(tabId, {
            type: "querySelector", selector, pierceShadow,
          });
          met = !resp?.result || resp.result.visible === false;
          break;
        }
        case "text_match": {
          if (!text) return { content: [{ type: "text", text: "text required for text_match" }] };
          await ensureAttached(tabId);
          const result = await cdp(tabId, "Runtime.evaluate", {
            expression: `document.body?.innerText || ""`,
            returnByValue: true,
          });
          const pageText = result?.result?.value || "";
          try {
            met = new RegExp(text, "i").test(pageText);
          } catch {
            met = pageText.includes(text);
          }
          break;
        }
        case "network_idle": {
          await sleep(1000);
          met = true; // simplified — wait 1s of no activity
          break;
        }
        case "js_predicate": {
          if (!predicate) return { content: [{ type: "text", text: "predicate required for js_predicate" }] };
          await ensureAttached(tabId);
          const result = await cdp(tabId, "Runtime.evaluate", {
            expression: predicate,
            returnByValue: true,
            awaitPromise: true,
          });
          met = !!result?.result?.value;
          break;
        }
        default:
          return { content: [{ type: "text", text: `Unknown condition: ${condition}` }] };
      }

      if (met) {
        const elapsed = ((Date.now() - start) / 1000).toFixed(1);
        return { content: [{ type: "text", text: `Condition "${condition}" met after ${elapsed}s` }] };
      }
      await sleep(pollInterval);
    }

    return { content: [{ type: "text", text: `Timeout: "${condition}" not met within ${timeout / 1000}s` }] };
  },

  async storage(args) {
    const { tabId, action, store = "localStorage", key, value } = args;
    if (!(await isInGroup(tabId))) return notInGroup(tabId);
    await ensureAttached(tabId);

    switch (action) {
      case "get": {
        if (!key) return { content: [{ type: "text", text: "key required" }] };
        const r = await cdp(tabId, "Runtime.evaluate", {
          expression: `${store}.getItem(${JSON.stringify(key)})`,
          returnByValue: true,
        });
        return { content: [{ type: "text", text: r?.result?.value ?? "null" }] };
      }
      case "set": {
        if (!key) return { content: [{ type: "text", text: "key required" }] };
        await cdp(tabId, "Runtime.evaluate", {
          expression: `${store}.setItem(${JSON.stringify(key)}, ${JSON.stringify(value || "")})`,
        });
        return { content: [{ type: "text", text: `Set ${store}["${key}"]` }] };
      }
      case "delete": {
        if (!key) return { content: [{ type: "text", text: "key required" }] };
        await cdp(tabId, "Runtime.evaluate", {
          expression: `${store}.removeItem(${JSON.stringify(key)})`,
        });
        return { content: [{ type: "text", text: `Deleted ${store}["${key}"]` }] };
      }
      case "list": {
        const r = await cdp(tabId, "Runtime.evaluate", {
          expression: `JSON.stringify(Object.keys(${store}))`,
          returnByValue: true,
        });
        return { content: [{ type: "text", text: r?.result?.value || "[]" }] };
      }
      case "get_cookies": {
        const tab = await chrome.tabs.get(tabId);
        const url = tab.url;
        const cookies = await chrome.cookies.getAll({ url });
        const text = cookies.map(c => `${c.name}=${c.value} (${c.domain})`).join("\n");
        return { content: [{ type: "text", text: text || "No cookies" }] };
      }
      default:
        return { content: [{ type: "text", text: `Unknown storage action: ${action}` }] };
    }
  },

  async dom_query(args) {
    const { tabId, selector, pierceShadow = true, includeStyles = false, limit = 20 } = args;
    if (!(await isInGroup(tabId))) return notInGroup(tabId);

    const resp = await sendContentMessage(tabId, {
      type: "domQuery", selector, pierceShadow, includeStyles, limit,
    });

    if (!resp?.result || resp.result.length === 0) {
      return { content: [{ type: "text", text: `No elements match "${selector}"` }] };
    }

    let text = `Found ${resp.result.length} element(s) for "${selector}":\n\n`;
    for (const el of resp.result) {
      text += `[${el.ref}] <${el.tag}> "${el.text}" at (${el.x}, ${el.y}) ${el.w}x${el.h}`;
      if (el.styles) text += ` styles: ${JSON.stringify(el.styles)}`;
      text += "\n";
    }
    return { content: [{ type: "text", text }] };
  },
};

File: src/mcp/tools_advanced.rs
use super::tools::{Tool, tab_prop};
use serde_json::json;

pub fn advanced_tools() -> Vec<Tool> {
    vec![
        Tool {
            name: "wait_for",
            description: "Wait for a condition: element visible/hidden, text match, network idle, or JS predicate.",
            input_schema: json!({
                "type": "object",
                "properties": {
                    "tabId": tab_prop(),
                    "condition": { "type": "string", "enum": ["element_visible","element_hidden","text_match","network_idle","js_predicate"] },
                    "selector": { "type": "string", "description": "CSS selector for element conditions." },
                    "text": { "type": "string", "description": "Text/regex for text_match." },
                    "predicate": { "type": "string", "description": "JS expression for js_predicate." },
                    "timeout": { "type": "number", "description": "Max seconds (default 10, max 30)." },
                    "pierceShadow": { "type": "boolean", "description": "Search Shadow DOM (default true)." }
                },
                "required": ["tabId", "condition"]
            }),
        },
        Tool {
            name: "storage",
            description: "Read/write localStorage, sessionStorage, cookies for the current page.",
            input_schema: json!({
                "type": "object",
                "properties": {
                    "tabId": tab_prop(),
                    "action": { "type": "string", "enum": ["get","set","delete","list","get_cookies"] },
                    "store": { "type": "string", "enum": ["localStorage","sessionStorage"] },
                    "key": { "type": "string" },
                    "value": { "type": "string" }
                },
                "required": ["tabId", "action"]
            }),
        },
        Tool {
            name: "dom_query",
            description: "CSS selector query with Shadow DOM piercing, computed styles, bounding rects.",
            input_schema: json!({
                "type": "object",
                "properties": {
                    "tabId": tab_prop(),
                    "selector": { "type": "string", "description": "CSS selector." },
                    "pierceShadow": { "type": "boolean", "description": "Traverse Shadow DOM (default true)." },
                    "includeStyles": { "type": "boolean", "description": "Include computed styles." },
                    "limit": { "type": "number", "description": "Max results (default 20)." }
                },
                "required": ["tabId", "selector"]
            }),
        },
    ]
}
